File: web/src/services/parcelMarkers.ts
import type { NsprdFeatureCollection } from "./nsprd";

/**
 * Listed parcels are sub-pixel polygons at overview zooms, so the map shows
 * circle markers instead while zoom <= this threshold. Polygon fills take
 * over from zoom 12, where parcel shapes become legible.
 */
export const OVERVIEW_MARKER_MAX_ZOOM = 11;

export type RepresentativePoint = {
  pid: string;
  latitude: number;
  longitude: number;
};

type Ring = readonly (readonly number[])[];

function ringAreaAndCentroid(ring: Ring): {
  area: number;
  latitude: number;
  longitude: number;
} {
  let twiceArea = 0;
  let cx = 0;
  let cy = 0;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i, i += 1) {
    const [x1, y1] = ring[j];
    const [x2, y2] = ring[i];
    const cross = x1 * y2 - x2 * y1;
    twiceArea += cross;
    cx += (x1 + x2) * cross;
    cy += (y1 + y2) * cross;
  }
  if (twiceArea === 0) {
    const [x, y] = ring[0];
    return { area: 0, longitude: x, latitude: y };
  }
  return {
    area: Math.abs(twiceArea / 2),
    longitude: cx / (3 * twiceArea),
    latitude: cy / (3 * twiceArea),
  };
}

/**
 * One marker point per listed PID: the centroid of the largest outer ring
 * among that PID's returned NSPRD features. Planar math on lat/lon is fine
 * here — parcels are far too small for projection error to move a marker
 * visibly at the zooms where markers show.
 */
export function representativeParcelPoints(
  parcels: NsprdFeatureCollection,
  pids: Set<string>,
): RepresentativePoint[] {
  const best = new Map<
    string,
    { area: number; latitude: number; longitude: number }
  >();
  for (const feature of parcels.features) {
    const pid = feature.properties?.PID;
    if (!pid || !pids.has(pid)) {
      continue;
    }
    const { geometry } = feature;
    const outerRings: Ring[] =
      geometry.type === "Polygon"
        ? [geometry.coordinates[0]]
        : geometry.type === "MultiPolygon"
          ? geometry.coordinates.map((polygon) => polygon[0])
          : [];
    for (const ring of outerRings) {
      if (!ring || ring.length < 4) {
        continue;
      }
      const candidate = ringAreaAndCentroid(ring);
      const current = best.get(pid);
      if (!current || candidate.area > current.area) {
        best.set(pid, candidate);
      }
    }
  }
  return [...best.entries()].map(([pid, point]) => ({
    pid,
    latitude: point.latitude,
    longitude: point.longitude,
  }));
}

File: NSMarksCore/Sources/GeoCore/ParcelMarkers.swift
import Foundation

/// Where a listed parcel is, when the parcel itself is too small to see.
///
/// Ported from `web/src/services/parcelMarkers.ts`. At an overview zoom a
/// parcel is a sub-pixel polygon: a user looking at the whole province would
/// see nothing at all where a tax sale is, which is the one thing that view is
/// for. A marker stands in until the boundary is legible.
public enum ParcelMarkers {
    /// Markers show at this zoom and below. Polygon fills take over from 12,
    /// where a parcel's shape becomes readable — the same threshold the web
    /// uses, so the two surfaces change over at the same place.
    public static let overviewMaxZoom = 11

    /// The centroid of the largest outer ring, or nil for a parcel with no
    /// boundary to stand for.
    ///
    /// Largest ring rather than all of them averaged: a parcel split by a road
    /// is several parts, and the mean of the parts can land on the road. The
    /// biggest piece is the one a marker over it is honestly pointing at.
    ///
    /// Planar maths on degrees, as on the web. A parcel is far too small for
    /// the projection error to move a marker visibly at the zooms where markers
    /// are drawn at all.
    public static func representativePoint(parts: [[[GeoPoint]]]) -> GeoPoint? {
        var best: (area: Double, point: GeoPoint)?
        for part in parts {
            guard let outer = part.first, outer.count >= 4 else { continue }
            let candidate = ringAreaAndCentroid(outer)
            if best == nil || candidate.area > best!.area {
                best = candidate
            }
        }
        return best?.point
    }

    private static func ringAreaAndCentroid(_ ring: [GeoPoint]) -> (area: Double, point: GeoPoint) {
        var twiceArea = 0.0
        var x = 0.0
        var y = 0.0
        var previous = ring.count - 1
        for index in ring.indices {
            let first = ring[previous]
            let second = ring[index]
            let cross = first.lng * second.lat - second.lng * first.lat
            twiceArea += cross
            x += (first.lng + second.lng) * cross
            y += (first.lat + second.lat) * cross
            previous = index
        }
        guard twiceArea != 0 else {
            // A ring of zero area is a degenerate boundary, not an error. Its
            // first vertex is as good a place to point at as any.
            return (0, ring[0])
        }
        return (
            abs(twiceArea / 2),
            GeoPoint(lat: y / (3 * twiceArea), lng: x / (3 * twiceArea))
        )
    }
}
